---
title: "sample usage"
output: pdf_document
---

```{r}
library(badger)
```

Expression-based analysis


Get common het SNPs from ExAC database.
```{r}
vcfFile <- "data-raw/ExAC.r0.3.sites.vep.vcf.gz"

# example with region of chromosome
chr <- 1
testRanges <- GRanges(chr, IRanges(start = 80000000, width=10000000))
param = ScanVcfParam(which=testRanges)
vcf <- readVcf(vcfFile, "hg19", param=param) 

# common snps by MAF
info <- as.data.frame(info(vcf))
print(dim(info))
print(head(info))
maf <- info[, 'AF'] # AF is Integer allele frequency for each Alt allele
print("number of snps with maf > 0.1:")
vi <- sapply(maf, function(x) any(x > 0.1))
print(table(vi))

# convert to alleleInfo
snpsDf <- as.data.frame(rowData(vcf)[vi,])
alleleInfo <- data.frame(
    'contig' = paste0('chr', as.character(snpsDf[,1])),
    'position' = as.numeric(snpsDf[,2]),
    'ref_allele' = as.character(snpsDf$REF),
    'alt_allele' = sapply(snpsDf$ALT, function(i) paste(as.character(i), collapse=',')),
    stringsAsFactors = FALSE
    )
alleleInfo <- cbind(alleleInfo, 'AF'=maf[vi])
    
# get rid of non single nucleotide changes
vi <- sapply(alleleInfo$ref_allele, nchar) == 1
alleleInfo <- alleleInfo[vi,]
# also gets rid of sites with multiple alt alleles though...hard to know which is in our patient
vi <- sapply(alleleInfo$alt_allele, nchar) == 1
alleleInfo <- alleleInfo[vi,]
# fix chromosome name
alleleInfo[,1] <- gsub('chr', '', alleleInfo[,1])

print(head(alleleInfo))
```

Now that we have putative heterozygous germline SNPs, we can get the coverage at these SNP sites from our bams. 

```{r}
print("Getting coverage...")
path <- 'data-raw/'
files <- list.files(path = path)
files <- files[grepl('.bam$', files)]    
cov <- do.call(cbind, lapply(files, function(f) {
    print(f)
    bamFile <- paste0(path, f)
    indexFile <- paste0(path, paste0(f, '.bai'))
    getCoverage(alleleInfo, bamFile, indexFile)    
}))
colnames(cov) <- files
head(cov)

# any coverage?
print("Snps with coverage:")
print(table(rowSums(cov)>0))
range(rowSums(cov>0))
vi <- rowSums(cov)>0
cov <- cov[vi,]

alleleInfo <- alleleInfo[vi,]

print("Getting allele counts...")
alleleCounts <- lapply(files, function(f) {
    print(f)
    bamFile <- paste0(path, f)
    indexFile <- paste0(path, paste0(f, '.bai'))
    getAlleleCount(alleleInfo, bamFile, indexFile)    
})
altCounts <- do.call(cbind, lapply(1:length(alleleCounts), function(i) alleleCounts[[i]][[1]]))
refCounts <- do.call(cbind, lapply(1:length(alleleCounts), function(i) alleleCounts[[i]][[2]]))
colnames(altCounts) <- colnames(refCounts) <- files

print(head(altCounts))
print(head(refCounts))
```

Now that we have these allele counts, we can plot

```{r}
r <- altCounts
n.sc <- altCounts+refCounts 
l <- rowSums(r>0)
n.bulk <- rowSums(n.sc>0)
clafProfile(r, n.sc, l, n.bulk)
```



